From b11b2dee1f38120a34802eb4a453e52faa338079 Mon Sep 17 00:00:00 2001
From: Justin Maggard <jmaggard@netgear.com>
Date: Wed, 7 Dec 2016 23:43:43 -0800
Subject: [PATCH] btrfs: don't automatically add a qgroup for snapshots

These serve no purpose in our use case, so just get rid of them.
---
 fs/btrfs/qgroup.c      |  9 +++++++++
 fs/btrfs/transaction.c | 11 +++++++----
 2 files changed, 16 insertions(+), 4 deletions(-)

diff --git a/fs/btrfs/qgroup.c b/fs/btrfs/qgroup.c
index c7c06c9..2a8ab5c 100644
--- a/fs/btrfs/qgroup.c
+++ b/fs/btrfs/qgroup.c
@@ -933,6 +933,14 @@ int btrfs_quota_enable(struct btrfs_trans_handle *trans,
 		btrfs_item_key_to_cpu(leaf, &found_key, slot);
 
 		if (found_key.type == BTRFS_ROOT_REF_KEY) {
+			struct btrfs_root *root;
+
+			key.objectid = found_key.offset;
+			key.type = BTRFS_ROOT_ITEM_KEY;
+			key.offset = (u64)-1;
+			root = btrfs_read_fs_root_no_name(fs_info, &key);
+			if (IS_ERR(root) || btrfs_root_readonly(root))
+				goto next_item;
 			ret = add_qgroup_item(trans, quota_root,
 					      found_key.offset);
 			if (ret)
@@ -944,6 +952,7 @@ int btrfs_quota_enable(struct btrfs_trans_handle *trans,
 				goto out_free_path;
 			}
 		}
+next_item:
 		ret = btrfs_next_item(tree_root, path);
 		if (ret < 0)
 			goto out_free_path;
diff --git a/fs/btrfs/transaction.c b/fs/btrfs/transaction.c
index 69314ad..1dd94ba 100644
--- a/fs/btrfs/transaction.c
+++ b/fs/btrfs/transaction.c
@@ -1598,10 +1598,13 @@ static noinline int create_pending_snapshot(struct btrfs_trans_handle *trans,
 	 * To co-operate with that hack, we do hack again.
 	 * Or snapshot will be greatly slowed down by a subtree qgroup rescan
 	 */
-	ret = qgroup_account_snapshot(trans, root, parent_root,
-				      pending->inherit, objectid);
-	if (ret < 0)
-		goto fail;
+	/* Don't automatically add a qgroup for read-only snapshots */
+	if (!(root_flags & BTRFS_ROOT_SUBVOL_RDONLY)) {
+		ret = qgroup_account_snapshot(trans, root, parent_root,
+					      pending->inherit, objectid);
+		if (ret < 0)
+			goto fail;
+	}
 
 	ret = btrfs_insert_dir_item(trans, parent_root,
 				    dentry->d_name.name, dentry->d_name.len,
-- 
1.9.1

